vernac/egramml.ml

```diff
@@ -31,13 +31,21 @@ type ('self, 'tr, _, 'r) ty_rule =
 type ('self, 'r) any_ty_rule =
 | AnyTyRule : ('self, _, 'act, Loc.t -> 'r) ty_rule -> ('self, 'r) any_ty_rule
 
+let keyword_needed = function
+  | GramTerminal _ :: _ -> false
+  | GramNonTerminal _ :: _ | [] -> true (* at level 0 of the grammar, it could be "false" *)
+
 let rec ty_rule_of_gram = function
 | [] -> AnyTyRule TyStop
 | GramTerminal s :: rem ->
+  let kw_need = keyword_needed rem in
   let AnyTyRule rem = ty_rule_of_gram rem in
-  let tok = Pcoq.Symbol.token (Pcoq.terminal s) in
-  let r = TyNext (rem, tok, None) in
-  AnyTyRule r
+  if kw_need then
```

Comment:
Same remark about factorization.

Reply:
I  added a comment to explain why there is no factorization.